Pegasus/include/Geometry.hpp

```diff
@@ -9,237 +9,251 @@
+
+    Cylinder();
+
     Cylinder(glm::dvec3 const& centerOfMass, glm::dvec3 const& halfHeight, double r);
 };
 
+/** Box data storage class */
 class Box : public SimpleShape
 {
 public:
-    Box(glm::dvec3 const& centerOfMass, glm::dvec3 const& a, glm::dvec3 const& b, glm::dvec3 const& c);
+    glm::dvec3 iAxis;
+    glm::dvec3 jAxis;
+    glm::dvec3 kAxis;
 
-    void SetAxes(glm::dvec3 const& a, glm::dvec3 const& b, glm::dvec3 const& c);
+    Box();
 
-    void GetAxes(glm::dvec3& a, glm::dvec3& b, glm::dvec3& c) const;
-
-private:
-    glm::dvec3 m_aAxis;
-    glm::dvec3 m_bAxis;
-    glm::dvec3 m_cAxis;
+    Box(glm::dvec3 const& centerOfMass, glm::dvec3 const& i, glm::dvec3 const& j, glm::dvec3 const& k);    
 };
 
 namespace intersection
 {
-// Utility functions
+/** Base cache data structure for shapes intersection queries */
+struct IntersectionCacheBase
+{
+};
+
+/** Specialized cache data structure for shapes intersection queries */
+template <typename ShapeA, typename ShapeB>
+struct IntersectionCache : IntersectionCacheBase
+{
+};
 
-/**
- * @brief Calculates box vertices in the model coordinate space from a given orthogonal basis
- *
- * Writes output vertices to the container starting with @p verticesBeginIterator. There must
- * be at least 7 more elements following given iterator.
- * @tparam Random access iterator
- * @param[in] i box axis vector
- * @param[in] j box axis vector
- * @param[in] k box axis vector
- * @param[in] verticesBeginIterator iterator to the container that is able to store 8 vertices
- */
-template <typename VerticesContainerIt>
-void CalculateBoxVertices(
-        glm::dvec3 const& i, glm::dvec3 const& j, glm::dvec3 const& k,
-        VerticesContainerIt verticesBeginIterator
-    )
-{
-    *(verticesBeginIterator + 0) = ( i + j + k);
-    *(verticesBeginIterator + 1) = ( i - j + k);
-    *(verticesBeginIterator + 2) = ( j - i + k);
-    *(verticesBeginIterator + 3) = (-i - j + k);
-    *(verticesBeginIterator + 4) = ( i + j - k);
-    *(verticesBeginIterator + 5) = ( i - j - k);
-    *(verticesBeginIterator + 6) = ( j - i - k);
-    *(verticesBeginIterator + 7) = (-i - j - k);
-}
-
-template <typename SrcIt1, typename SrcIt2, typename DestIt>
-void CalculateSeparatingAxes(SrcIt1 srcBegin1, SrcIt1 srcEnd1, SrcIt2 srcBegin2, SrcIt2 srcEnd2,
-    std::back_insert_iterator<DestIt> destBegin)
-{
-    for (auto it1 = srcBegin1; it1 != srcEnd1; ++it1)
-    {
-        for (auto it2 = srcBegin2; it2 != srcEnd2; ++it2)
-        {
-            auto const axis = glm::normalize(glm::cross(*it1, *it2));
-            if (glm::length2(axis) != 0.0)
-            {
-                destBegin++ = axis;
-            }
-        }
-    }
-}
+template <>
+struct IntersectionCache<Ray, Ray> : IntersectionCacheBase
+{
+    glm::dvec3 aRayNormal;
+    glm::dvec3 aRayPoint;
+    glm::dvec3 bRayNormal;
+    glm::dvec3 bRayPoint;
+    double denominator;
+    glm::dvec3 aClosestApproach;
+    glm::dvec3 bClosestApproach;
```

Comment:
Don't see a reason to construct those, frankly. Clarity is not suffering from this structure of storing.